feat: support deposit and withdraw transaction

shayanh opened the pull request:
This PR mainly adds support for:
- L1-initiated transactions
- Deposit transactions
- Withdraw transaction

Additionally, it includes critical bug fixes and improvements across the codebase. Perhaps would have been if I had broken it into several PRs.

---

## Features

### Deposit Transaction

Added `DepositTransactionClient` similar to the existing update transaction client.

```typescript
// Create a deposit transaction
const depositTx = await createDepositTransactionClient({
  keystoreAddress: userAcct.address,
  amt: parseEther("0.01"),
});
```

### Withdraw Transaction

Added `WithdrawTransactionClient` similar to the existing update transaction client.

```typescript
const withtrawTx = await createWithdrawTransactionClient({
  amt: parseEther("0.005"),
  to: account.address,
  userAcct,
  nodeClientUrl: config.l2RpcUrl,
  sequencerClientUrl: config.l2RpcUrl,
});
```

### L1-initiation Flow

Added `publicActionsL1` and `walletActionsL1` viem L1-client extensions to facilitate initiation of L1 transactions. This design is inspired by op-stack's viem integration.

```typescript
const l1Client = createWalletClient({
  account,
  transport: http(config.l1RpcUrl),
})
  .extend(publicActionsL1())
  .extend(walletActionsL1());

// Send the deposit transaction to L1
const l1TxHash = await l1Client.initiateL1Transaction({
  bridgeAddress: config.bridgeAddress,
  txClient: depositTx,
});
console.log("L1 transaction hash:", l1TxHash);

// Fetch deposit transaction hash and receipt
const l1TxReceipt = await l1Client.waitForTransactionReceipt({ hash: l1TxHash });

const [l2TxHash] = getL2TransactionHashes(l1TxReceipt);
console.log("Deposit transaction hash:", l2TxHash);

const l2TxReceipt = await l2Client.waitForTransactionReceipt({ hash: l2TxHash });
console.log("Deposit transaction receipt:", l2TxReceipt);
```

## Examples

Renamed `example` folder to `examples` and added new `deposit-withdraw` example.

## Fixes and Improvements

- Fix `toBytes` and `txHash` methods to match our Rust implementation.
- Rename `signedTx` to `txSignature` to avoid confusion.
- Add `waitForTransactionReceipt` and `waitForTransactionFinalization` methods to `NodeClient`.
- Remove the misleading `waitForTransactionInclusion` method from `SequencerClient`.
- Signature prover:
  - Add `MOfNSignatureProver` to `signature-provers` module
  - Move `url` to `SignatureProverClientConfig` from `CustomSignatureProver`
  - Update signature prover section of the readme

Closes INT-3789, INT-3835

HrikB commented on README.md:
````diff
@@ -92,21 +107,21 @@ const updateTx = await createUpdateTransactionClient({
 ```
 
-We'll then go ahead and sign the transaction with our account's private key:
+### Authenticating a Transaction
````
```suggestion
### Authenticating an L2 Transaction
```

Since this doesn't apply to deposits.

HrikB reviewed (commented):
Overall looks good, one small comment on tx hash in deposit client and nits.

Also, two updates I realized we should make before release: 
- We need to update [this](https://github.com/axiom-crypto/axiom-keystore-sdk/blob/shayanh/l1-init/README.md?plain=1#L15-L26) part of the README since signature prover client is no longer an external repo.
- Do we want to update bridge address [here](https://github.com/axiom-crypto/axiom-keystore-sdk/blob/shayanh/l1-init/src/constants.ts#L6)?

shayanh commented on examples/deposit-withdraw/src/index.ts:
```diff
@@ -0,0 +1,147 @@
+import "dotenv/config";
+
+// Codehash for EOA
+const EOA_CODEHASH = "0x595b7552e60f6430c898abc2b292aa805e94834a576f57969406940f6d12d4d9";
```
What codehash is this? I thought it should be the codehash for EOAs:
> For externally owned accounts, the codeHash field is the hash of an empty string.

from: https://ethereum.org/en/developers/docs/accounts/

HrikB replied:
I don't see where this value is coming from. But looks like it's used to [construct](https://github.com/axiom-crypto/axiom-keystore-sdk/blob/shayanh/l1-init/examples/deposit-withdraw/src/index.ts#L79) key data meaning it's key data consumer code hash

shayanh replied:
Should I update sponsor codehash too?
https://github.com/axiom-crypto/axiom-keystore-sdk/blob/main/example/src/index.ts#L29

HrikB replied:
I think this is actually related to whatever account we're initializing as sponsor in the backend. My guess is this is being used for some counterfactual initialization.

The sponsor account never really has a non-keystore-L2 transaction flow so this field has no meaning to it, other than contributing to counterfactual init.

HrikB commented on src/transaction/deposit.ts:
```diff
@@ -0,0 +1,47 @@
+export async function createDepositTransactionClient(
+  tx: DepositTransactionInputs,
+): Promise<DepositTransactionClient> {
+  const l1InitiatedNonce = tx.l1InitiatedNonce
+    ? numberToHex(tx.l1InitiatedNonce, { size: 32 })
+    : pad("0x", { size: 32 });
+
+  const amt = tx.amt;
+
+  const keystoreAddress = tx.keystoreAddress;
+
+  const toBytes = (): Data => {
+    return encodePacked(
+      ["bytes1", "bytes", "uint256", "bytes"],
+      [TransactionType.Deposit, l1InitiatedNonce, amt, keystoreAddress],
+    );
+  };
+
+  const txHash = (): Hash => keccak256(toBytes());
```
Ok we can just leave it be for now then.

shayanh commented:
> We need to update [this](https://github.com/axiom-crypto/axiom-keystore-sdk/blob/shayanh/l1-init/README.md?plain=1#L15-L26) part of the README since signature prover client is no longer an external repo.

Signature prover needed some extra work, fixed here: 991e0a98bc3dc04dd66c8d0fb3b6d627414ed18d

> Do we want to update bridge address [here](https://github.com/axiom-crypto/axiom-keystore-sdk/blob/shayanh/l1-init/src/constants.ts#L6)?

Updated here: 3cc292ea7cebd94e5158a869895c0068aba52b93

shayanh commented:
Holding off the merge till prod deployment is fixed.
